zephyr/port/bootloader/bootloader_flash.c

```diff
@@ -333,53 +340,93 @@ static esp_err_t bootloader_flash_read_allow_decrypt(size_t src_addr, void *dest
 
 esp_err_t esp_rom_flash_read(size_t src_addr, void *dest, size_t size, bool allow_decrypt)
 {
-    if (src_addr & 3) {
-        ESP_EARLY_LOGE(TAG, "bootloader_flash_read src_addr 0x%x not 4-byte aligned", src_addr);
-        return ESP_FAIL;
-    }
-    if (size & 3) {
-        ESP_EARLY_LOGE(TAG, "bootloader_flash_read size 0x%x not 4-byte aligned", size);
-        return ESP_FAIL;
-    }
-    if ((intptr_t)dest & 3) {
-        ESP_EARLY_LOGE(TAG, "bootloader_flash_read dest 0x%x not 4-byte aligned", (intptr_t)dest);
-        return ESP_FAIL;
-    }
+    uint8_t *dest_ptr = (uint8_t *)dest;
+    size_t aligned_size, copy_size;
+    size_t remaining = size;
+    uint8_t temp_buf[FLASH_BLK_SZ];
 
-    if (allow_decrypt) {
-        return bootloader_flash_read_allow_decrypt(src_addr, dest, size);
-    } else {
-        return bootloader_flash_read_no_decrypt(src_addr, dest, size);
+    /* handle misaligned source address */
+    size_t src_offset = src_addr & 3;
+    size_t start_addr = src_addr - src_offset;
+
+    while (remaining > 0) {
+        copy_size = (remaining > FLASH_BLK_SZ) ? FLASH_BLK_SZ : remaining;
+        aligned_size = (copy_size + src_offset + 3) & ~3;
+
+        /* read from flash into the temp buffer */
+        esp_err_t err;
+        if (allow_decrypt) {
+            err = bootloader_flash_read_allow_decrypt(start_addr, temp_buf, aligned_size);
```

Comment:
`aligned_size` could be more than the buffer size, leading to buffer overflow.

Example (in first run of the loop):
```
src_addr = 3;
src_offset = 3;
start_addr = 0;
copy_size = 32;
aligned_size = (32 + 3 + 3) & ~3 = 36;
```

Reply:
Perhaps we can add the max offset which would be 3.. then rouding to 4:
`uint8_t temp_buf[FLASH_BLK_SZ + 4];`